File: algorithms/algorithms-cplusplus/leetcode/Question_0486_Predict_the_Winner.cc
#include <vector>

using namespace std;

class Solution {
 public:
  bool PredictTheWinner(vector<int>& nums) {
    int n = nums.size();
    vector<vector<int>> dp(n, vector<int>(n, -1));
    return canWin(nums, 0, n - 1, dp) >= 0;
  }

  int canWin(vector<int>& nums, int s, int e, vector<vector<int>>& dp) {
    if (dp[s][e] == -1) {
      dp[s][e] = (s == e) ? nums[s] : max(nums[s] - canWin(nums, s + 1, e, dp), nums[e] - canWin(nums, s, e - 1, dp));
    }
    return dp[s][e];
  }

  bool PredictTheWinner2(vector<int>& nums) {
    int n = nums.size();
    vector<vector<int>> dp(n, vector<int>(n, 0));
    for (int i = 0; i < n; ++i) {
      dp[i][i] = nums[i];
    }
    for (int len = 1; len < n; ++len) {
      for (int i = 0, j = len; j < n; ++i, ++j) {
        dp[i][j] = max(nums[i] - dp[i + 1][j], nums[j] - dp[i][j - 1]);
      }
    }
    return dp[0][n - 1] >= 0;
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0486_Predict_the_Winner.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.Arrays;

public class Question_0486_Predict_the_Winner {
    public boolean PredictTheWinner(int[] nums) {
        if (nums == null || nums.length == 0) {
            return false;
        }
        int n = nums.length;
        int[][] dp = new int[n][n];
        for (int i = 0; i < n; i++) {
            Arrays.fill(dp[i], -1);
        }
        return help(nums, 0, n - 1, dp) >= 0;
    }

    private int help(int[] nums, int s, int e, int[][] dp) {
        if (dp[s][e] == -1) {
            dp[s][e] = (s == e) ? nums[s] : Math.max(nums[s] - help(nums, s + 1, e, dp), nums[e] - help(nums, s, e - 1, dp));
        }
        return dp[s][e];
    }
}
